### PR TITLE
[Veeam] Force ACS to use the template repository to clone job

### PR DESCRIPTION
**Pull request by SadiJr**
### Description

Using the VMWare hypervisor, and with Veeam integration enabled, when a VM is assigned to a Backup Offering, ACS executes an operation to clone the Veeam job of that Backup Offering. However, instead of using the Backup Offering specific repository, currently ACS uses the first repository found on Veeam, which can cause problems for users who have more than one repository.

This PR intends to fix this behavior.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
It was tested in a local laboratory, with 2 Veeam repositories:

- I assigned two VMs to two Backup Offerings, one for each repository;
- The two jobs were created in the same repository, and it should be one for each. This happened because ACS is using the first repository it finds;
- With this changes, I repeated the test in the same environment;
- Now the jobs are being cloned in the correctly repositories.

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
cltgm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by GutoVeronezi — APPROVED** [review]
LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
@SadiJr do you want this on 4.16.1? (if so please rebase on the 4.16 branch)

**Comment by SadiJr** [comment]
@DaanHoogland For me, it is fine to go into main only.

File: plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java [hunk 1]
```diff
@@ -439,7 +461,12 @@ public boolean startBackupJob(final String jobId) {
     public boolean cloneVeeamJob(final Job parentJob, final String clonedJobName) {
         LOG.debug("Trying to clone veeam job: " + parentJob.getUid() + " with backup uuid: " + clonedJobName);
         try {
-            final Ref repositoryRef =  listBackupRepository(parentJob.getBackupServerId());
+            final Ref repositoryRef = listBackupRepository(parentJob.getBackupServerId(), parentJob.getName());
+            if (repositoryRef == null) {
+                throw new CloudRuntimeException(String.format("Failed to clone backup job because couldn't find any "
+                        + "repository associated with backup job [id: %s, uid: %s, backupServerId: %s, name: %s].",
+                        parentJob.getId(), parentJob.getBackupServerId(), parentJob.getUid(), parentJob.getName()));
```
Comment (sureshanaparti) [inline plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java:468]: @SadiJr correct the order of parameters in the error message.

```suggestion
                        parentJob.getId(), parentJob.getUid(), parentJob.getBackupServerId(), parentJob.getName()));
```

**Comment by SadiJr** [comment]
@sureshanaparti thanks for the review, and sorry for the delay in answer.

File: plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java [hunk 2]
```diff
@@ -337,25 +337,47 @@ private boolean checkTaskStatus(final HttpResponse response) throws IOException
...
+        );
+        Pair<Boolean, String> result = executePowerShellCommands(cmds);
+        if (result == null || !result.first()) {
+            throw new CloudRuntimeException(String.format("Failed to get Repository Name from Job [name: %s].", backupName));
+        }
+
+        for (String block : result.second().split("\n\n")) {
+           if (block.matches("Name(\\s)+:(.)*")) {
+               return block.split(":")[1].trim();
+           }
```
Comment (GabrielBrascher) [inline plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java:376]: Maybe a few unit test cases would be good, just to ensure this method sends the expected `cmds` (lines 364 - 367) and retrieves the correct String when processing the result (373 - 377).

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
not sure if any smoke tests are available that would be touched by this chance. (i don't think there are) I think we must trust the authors testing, @rohityadavcloud @GabrielBrascher @sureshanaparti ? (cc @SadiJr )

**Comment by GabrielBrascher** [comment]
@DaanHoogland I think that there are no smoke tests covering it indeed.
The code looks good but I have no VMware staging environment so I would not be able to test it. I am trusting in the author's testing it.

To mitigate issues or future changes in the code that could affect this behavior I`ve suggested extracting a few parts of the added method and creating unit tests to cover a few of the String parsing logic.

**Comment by sureshanaparti** [comment]
Hi @SadiJr Can you address the outstanding comments. Thanks.

**Comment by SadiJr** [comment]
@GabrielBrascher @sureshanaparti Sorry for the delay in answer. 

Regarding @GabrielBrascher's suggestions: thank you for them, I will apply them soon.

File: plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java [hunk 3]
```diff
@@ -337,25 +337,47 @@ private boolean checkTaskStatus(final HttpResponse response) throws IOException
...
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (SadiJr) [reply plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java]: Done.

File: plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java (hunk 2 above)
Comment (SadiJr) [reply plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java:373]: Done too. Thanks for the suggestions.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3 keepEnv

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Review by sureshanaparti — APPROVED** [review]

**Comment by sureshanaparti** [comment]
Merging based on approvals, smoke test results and author's tests.

**Pull request merged**